include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,499 @@
+            {
+                newPseudoMantissa >>= 1;
+                ++exp;
+            }
+            newPseudoMantissa &= (ieee754::traits<float64_t>::mantissaMask);
+
+            return createPreserveBitPattern(impl::assembleFloat64(sign, uint64_t(exp) << ieee754::traits<float64_t>::mantissaBitCnt, newPseudoMantissa));
+        }
+
+        emulated_float64_t operator/(const emulated_float64_t rhs)
+        {
+            // TODO: maybe add function to extract real mantissa
+            const uint64_t lhsRealMantissa = (ieee754::extractMantissa(data) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt));
+            const uint64_t rhsRealMantissa = ieee754::extractMantissa(rhs.data) | (1ull << ieee754::traits<float64_t>::mantissaBitCnt);
+            
+            const uint64_t sign = (data ^ rhs.data) & ieee754::traits<float64_t>::signMask;
+            int exp = ieee754::extractExponent(data) - ieee754::extractExponent(rhs.data) + ieee754::traits<float64_t>::exponentBias;
+
+            uint64_t2 lhsMantissaShifted = impl::shiftMantissaLeftBy53(lhsRealMantissa);
+            uint64_t mantissa = impl::divmod128by64(lhsMantissaShifted.x, lhsMantissaShifted.y, rhsRealMantissa);
+
+            if (mantissa & (0x1ull << 53))
+            {
+                ++exp;
+            }
+            else
+            {
+                mantissa >>= 1;
+            }
+
+            mantissa &= ieee754::traits<float64_t>::mantissaMask;
+
+            return createPreserveBitPattern(impl::assembleFloat64(sign, uint64_t(exp) << ieee754::traits<float64_t>::mantissaBitCnt, mantissa));
+        }
+
+        // relational operators
+        bool operator==(emulated_float64_t rhs)
+        {
+            if (!FastMath && (nbl::hlsl::isnan<uint64_t>(data) || nbl::hlsl::isnan<uint64_t>(rhs.data)))
+                return false;
+            // TODO: i'm not sure about this one
+            if (!FastMath && impl::areBothZero(data, rhs.data))
+                return true;
+
+            const emulated_float64_t xored = emulated_float64_t::createPreserveBitPattern(data ^ rhs.data);
+            // TODO: check what fast math returns for -0 == 0
+            if ((xored.data & 0x7FFFFFFFFFFFFFFFull) == 0ull)
+                return true;
+
+            return !(xored.data);
+        }
+        bool operator!=(emulated_float64_t rhs)
+        {
+            if (!FastMath && (nbl::hlsl::isnan<uint64_t>(data) || nbl::hlsl::isnan<uint64_t>(rhs.data)))
+                return true;
+            // TODO: i'm not sure about this one
+            if (!FastMath && impl::areBothSameSignZero(data, rhs.data))
+                return false;
+
+            const emulated_float64_t xored = emulated_float64_t::createPreserveBitPattern(data ^ rhs.data);
+            // TODO: check what fast math returns for -0 == 0
+            if ((xored.data & 0x7FFFFFFFFFFFFFFFull) == 0ull)
+                return false;
+
+            return xored.data;

```

Comment:
any reason for not doing `return !operator==(rhs);` ?

Reply:
no reason